### DISCUSSION TITLE
QImage from VipsImage?

### DISCUSSION THREAD

**Discussion by bkocev** [libvips/libvips#3829] [General]
Dear all, 

I am loading a large VipsImage and then resizing it and in end I want to save it as a QImage, but I am facing problems with the function vips_image_write_to_memory. It takes really a lot of time (several minutes). The resized image is 8000 X 5560.

VipsImage* vipsImageResized;
vips_resize( vipsImage, &vipsImageResized, 1 / factor_, nullptr );
const uchar* data = static_cast< const uchar* >( vips_image_write_to_memory( vipsImageResized, nullptr ) );
img = QImage( std::move(data),
                          vips_image_get_width( vipsImageResized ),
                          vips_image_get_height( vipsImageResized ),
                          VIPS_IMAGE_SIZEOF_LINE( vipsImageResized ),
                          bands == 3 ? QImage::Format_RGB888 : QImage::Format_RGBA8888 );

Is there an alternative/faster way of constructing a QImage from a VipsImage?

Thank you in advance and kind regards,
Bojan.

**Comment by jcupitt** [comment]
Hello @bkocev,

The fastest path is usually with `thumbnail`, eg.:

```C++
VImage image = VImage::thumbnail("some-filename", 8000);
void* data = image.write_to_memory(nullptr); 
```

It does load + resize in one operation so it can use a range of load library tricks to speed up the resize. See:

https://www.libvips.org/API/current/cpp/classVImage.html#a7c0b607ff6c4ff700c02799d6c80abe3

And the full docs:

https://www.libvips.org/API/current/libvips-resample.html#vips-thumbnail

> img = QImage( std::move(data),

Are you sure you need this? You should be able to steal the memory from libvips.

Ultimately though you are limited by the speed of the image decode library, and there's only so much that's possible.

Is this for display to the user? Another option is to load in the background, driven by display expose events, This is what eg. `vipsdisp` does:

https://github.com/jcupitt/vipsdisp

It might be worth timing that with a test file. There's a binary on flathub:

https://flathub.org/apps/org.libvips.vipsdisp

**Reply by bkocev** [reply]
Hello,

the initial suggestion fixed the problem. I use vips_thumbnail_buffer and vips_thumbnail from the C api. It decreased the loading and rescaling from 3 min 12 sec to 25 sec. This is acceptable. Yes, it is for a display to the user.

Best,
Bojan.
